src/Lean/Compiler/IR/EmitLLVM.lean

```diff
@@ -0,0 +1,1521 @@
+
+namespace Lean.IR
+
+def leanMainFn := "_lean_main"
+
+namespace LLVM
+-- TODO(bollu): instantiate target triple and find out what size_t is.
+def size_tType (llvmctx : LLVM.Context) : IO (LLVM.LLVMType llvmctx) :=
+  LLVM.i64Type llvmctx
+end LLVM
+
+namespace EmitLLVM
+
+structure Context (llvmctx : LLVM.Context) where
+  env        : Environment
+  modName    : Name
+  jpMap      : JPParamsMap := {}
+  mainFn     : FunId := default
+  mainParams : Array Param := #[]
+  llvmmodule : LLVM.Module llvmctx
+
+structure State (llvmctx : LLVM.Context) where
+  var2val : HashMap VarId (LLVM.LLVMType llvmctx × LLVM.Value llvmctx)
+  jp2bb   : HashMap JoinPointId (LLVM.BasicBlock llvmctx)
+
+inductive Error where
+| unknownDeclaration : Name → Error
+| invalidExportName : Name → Error
+| unimplemented : String → Error
+| compileError : String → Error -- TODO: change this into an exhaustive Inductive enumeration.
+
+instance : ToString Error where
+  toString e := match e with
+   | .unknownDeclaration n => s!"unknown declaration '{n}'"
+   | .invalidExportName n => s!"invalid export name '{n}'"
+   | .unimplemented s => s!"unimplemented '{s}'"
+   | .compileError s => s!"compile error '{s}'"
+
+abbrev M (llvmctx : LLVM.Context) :=
+  StateT (State llvmctx) (ReaderT (Context llvmctx) (ExceptT Error IO))
+
+instance : Inhabited (M llvmctx α) where
+  default := throw (Error.compileError "inhabitant")
+
+def addVartoState (x : VarId) (v : LLVM.Value llvmctx) (ty : LLVM.LLVMType llvmctx) : M llvmctx Unit := do
+  modify (fun s => { s with var2val := s.var2val.insert x (ty, v) }) -- add new variable
+
+def addJpTostate (jp : JoinPointId) (bb : LLVM.BasicBlock llvmctx) : M llvmctx Unit :=
+  modify (fun s => { s with jp2bb := s.jp2bb.insert jp bb })
+
+def emitJp (jp : JoinPointId) : M llvmctx (LLVM.BasicBlock llvmctx) := do
+  let state ← get
+  match state.jp2bb.find? jp with
+  | .some bb => return bb
+  | .none => throw (Error.compileError s!"unable to find join point {jp}")
+
+def getLLVMModule : M llvmctx (LLVM.Module llvmctx) := Context.llvmmodule <$> read
+
+def getEnv : M llvmctx Environment := Context.env <$> read
+
+def getModName : M llvmctx  Name := Context.modName <$> read
+
+def getDecl (n : Name) : M llvmctx Decl := do
+  let env ← getEnv
+  match findEnvDecl env n with
+  | some d => pure d
+  | none   => IO.eprintln "getDecl failed!"; throw (Error.unknownDeclaration n)
+
+def constIntUnsigned (n : Nat) : M llvmctx (LLVM.Value llvmctx) :=  do
+    LLVM.constIntUnsigned llvmctx (UInt64.ofNat n)
+
+def getOrCreateFunctionPrototype (mod : LLVM.Module llvmctx)
+  (retty : LLVM.LLVMType llvmctx) (name : String) (args : Array (LLVM.LLVMType llvmctx)) : M llvmctx  (LLVM.Value llvmctx) := do
+  LLVM.getOrAddFunction mod name $
+     (← LLVM.functionType retty args (isVarArg := false))
+
+def callLeanBox (builder : LLVM.Builder llvmctx) (arg : LLVM.Value llvmctx) (name : String := "") : M llvmctx (LLVM.Value llvmctx) := do
```

Comment:
It would be pretty cool if we had some sort of general framework for this type of stuff. That is, making a Lean function of this style that automatically generates a call to an LLVM function with a given signature but I guess that's far out of scope here.

Reply:
Agreed, it would be nice!